src/lib/ngc/ngcc-processor.ts

```diff
@@ -0,0 +1,114 @@
+import { Logger, PathMappings } from '@angular/compiler-cli/ngcc';
+import { existsSync } from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+import * as log from '../util/log';
+import { EntryPointNode, ngUrl } from '../ng-v5/nodes';
+
+// Transform a package and it's typings when NGTSC is resolving a module.
+export class NgccProcessor {
+  private _processedModules = new Set<string>();
+  private _logger: NgccLogger;
+  private _nodeModulesDirectory: string;
+  private _ngcc: typeof import('@angular/compiler-cli/ngcc');
+  private _pathMappings: PathMappings | undefined;
+  private _entryPointsUrl: string[];
+
+  constructor(private readonly compilerOptions: ts.CompilerOptions, private readonly entryPoints: EntryPointNode[]) {
+    this._ngcc = require('@angular/compiler-cli/ngcc');
+    this._entryPointsUrl = this.entryPoints.map(({ url }) => ngUrl(url));
+
+    const { baseUrl, paths } = this.compilerOptions;
+    this._nodeModulesDirectory = this.findNodeModulesDirectory(baseUrl);
+
+    if (baseUrl && paths) {
+      this._pathMappings = {
+        baseUrl,
+        paths,
+      };
+    }
+  }
+
+  processModule(moduleName: string, resolvedModule: ts.ResolvedModule | ts.ResolvedTypeReferenceDirective): void {
+    const resolvedFileName = resolvedModule.resolvedFileName;
+    if (
+      !resolvedFileName ||
+      moduleName.startsWith('.') ||
+      this._processedModules.has(moduleName) ||
+      this._entryPointsUrl.includes(ngUrl(moduleName))
+    ) {
+      // Skip when module is unknown, relative or NGCC compiler is not found or already processed.
+      return;
+    }
+
+    const packageJsonPath = this.tryResolvePackage(moduleName, resolvedFileName);
+    if (!packageJsonPath) {
+      // add it to processed so the second time round we skip this.
+      this._processedModules.add(moduleName);
+
+      return;
+    }
+
+    this._ngcc.process({
+      basePath: this._nodeModulesDirectory,
+      targetEntryPointPath: path.dirname(packageJsonPath),
+      compileAllFormats: false,
+      propertiesToConsider: ['fesm2015', 'fesm5', 'es2015', 'esm2015', 'esm5', 'main', 'module'],
+      createNewEntryPointFormats: true,
```

Comment:
If we have this, ngcc generates its output in a separate subdirectory `__ivy_ngcc__` and the `package.json` file is updated with new entries for properties like `main`, to become suffixed with `_ivy_ngcc` targeting the `__ivy_ngcc__` subdirectory. I don't know how this is dealt with exactly, but I believe the CLI is responsible for switching over to these properties somehow, instead of the original. How is this dealt with in ng-packagr?

Reply:
In ng-packagr we only care about the typings for type checking, these are not placed under `__ivy_ngcc`.

Unfortunately ngcc doesn’t offer a way to process typings yet. That being said that is not a big deal. As the other module formats will probable be needed when running unit tests.

Reply:
Ah, that makes total sense. I see now, the `.d.ts` files are the new `metadata.json` :-)